### PR TITLE
executor: recovery panic in parallel hashagg and projection

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

prevent tidb-server from crash when panic happened(e.g. #8118) below parallel hashagg and projection operator.

### What is changed and how it works?

add `recover` to panic-able  goroutine, and send error result to user.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test modify code for panic and tidb-server doesn't crash.

Code changes

 - Add `recovery`

Side effects

 - N/A

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8185)

### CODE REVIEW

File: executor/aggregate.go [hunk 1]
```diff
@@ -314,9 +316,18 @@ func (w *HashAggPartialWorker) getChildInput() bool {
 	return true
 }
 
+func recoveryHashAgg(output chan *AfFinalResult, r interface{}) {
+	output <- &AfFinalResult{err: errors.Errorf("%v", r)}
+	buf := util.GetStack()
+	log.Errorf("panic in the recoverable goroutine: %v, stack trace:\n%s", r, buf)
+}
+
 func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup, finalConcurrency int) {
```
Comment (zz-jason) [inline executor/aggregate.go]: It's better to use `util.WithRecovery()`

Comment (lysu) [reply executor/aggregate.go]: I tried, but for this two operator using `util.WithRecovery()` is not suitable, except do recovery, we need take param `waitGroup` or even local variable `output` or `needShuffle` to do some cleanup operation(and recovery send response must be done before those cleanup), use withRecovery's closure to catch those variables will make more complex code, I think make reuseable `recoveryXXX` to recovery and send result is more simple and clear to this snippet.

Comment (eurekaka) [reply executor/aggregate.go]: Can we put the cleanup into a `recoverFn` of `WithRecovery`? like `HashJoinExec::finishOuterFetcher`?

Comment (lysu) [reply executor/aggregate.go]: @eurekaka `HashJoinExec::finishOuterFetcher` only depend on `HashJoinExec`'s fields, but here some local vairable like `needShuffle` or `projectionOutput` is local variable, we need modify more to share them to `recoveFn`.

**Comment by zz-jason** [comment]
@lysu we should cherrypick this patch to release-2.1 and 2.0?

**Pull request merged**